src/useEditor.ts

```diff
@@ -33,11 +44,56 @@ export type CreateView = (
   dispatch: (tr: Transaction) => EditorState
 ) => EditorView
 
-const useEditor = (initialState: EditorState, createView: CreateView) => {
+const useEditor = (
+  initialState: EditorState,
+  createView: CreateView,
+  editorProps: EditorProps
+) => {
   const view = useRef<EditorView>()
   const [state, setState] = useState<EditorState>(initialState)
   const [viewElement, setViewElement] = useState<HTMLDivElement | null>(null)
   const history = useHistory()
+  const { collabProvider } = editorProps
+
+  // Receiving steps from backend
+  if (collabProvider) {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    collabProvider.onNewSteps(async (newVersion, steps, clientIDs) => {
+      if (state && view.current) {
+        // @TODO: make sure received steps are ignored by the quarterback plugin
+        const localVersion = getVersion(view.current.state)
+
+        // @TODO - save unconfirmed verison and compare it with newVersion to check if we can consume this update and don't have to call collabProvider.stepsSince
+        // if (newVersion == lastLocalUnconfirmed) {
+        //   view.current.dispatch(
+        //     receiveTransaction(
+        //       // has to be called for the collab to increment version and drop buffered steps
+        //       view.current.state,
+        //       steps,
+        //       clientIDs
+        //     )
+        //   )
+        // }
+
+        const since = await collabProvider.stepsSince(localVersion)
+
+        if (since?.steps && since.clientIDs) {
+          view.current.dispatch(
+            receiveTransaction(
+              // has to be called for the collab to increment version and drop buffered steps
+              view.current.state,
+              since?.steps,
+              since.clientIDs
+            )
+          )
+        } else {
+          console.warn('Inconsistent new steps event from the authority.')
+        }
+      }
+    })
+  }
```

Comment:
onNewSteps can only have a single listener, so there is no risk of duplication for that callback. State changes everytime anyway so even if we'll wrap it with a useEffect it will be recreated on each rerender anyway.

This is the onNewSteps dispatcher implementation:
![image](https://github.com/Atypon-OpenSource/manuscripts-body-editor/assets/40764175/6f491537-920b-48d0-89bb-113327da2c3f)